### DISCUSSION TITLE
Allow user to reorder list?

### DISCUSSION THREAD

**Discussion by AnderssonPeter** [tui-cs/Terminal.Gui#1490] [Ideas]
Is there someway to use this framework where i can provide a list of items and allow the user to change the order of them?

**Comment by BDisp** [comment]
Of course you can. Just sort the source list the way you want and set the list source again.

**Comment by tznind** [comment]
Here is an example

```csharp
Application.Init ();
var lv = new ListView (new List<string> { "dog", "cat", "horse", "catipillar" }) {
	Width = 10,
	Height = 10
};

var btnAsc = new Button (0, 11, "Asc");
btnAsc.Clicked += () => {
	lv.SetSource(lv.Source.ToList ().Cast<string> ().OrderBy (a => a).ToList ());
};
var btnDesc = new Button (0, 12, "Desc");
btnDesc.Clicked += () => {
	lv.SetSource (lv.Source.ToList ().Cast<string> ().OrderByDescending (a => a).ToList ());
};

var win = new Window ();
win.Add (lv);
win.Add (btnAsc);
win.Add (btnDesc);

Application.Run (win);
Application.Shutdown ();
```

![image](https://user-images.githubusercontent.com/31306100/137028945-7023a2de-3f85-4e1c-9c37-980533fd0760.png)

**Comment by AnderssonPeter** [comment]
What im trying to achieve is allowing the user to move a specific item up or down the list, exampe we have the items A, B, C.
The user scrolls down to B selects it and pressed Up, now the list should be B, A, C.

Bonus if the item chages background color while selected.

**Comment by BDisp** [comment]
See here:
https://github.com/migueldeicaza/gui.cs/pull/1489

In the `DynamicMenuBar` and in the `DynamicStatusBar` there is something for what you want.

![menubar](https://user-images.githubusercontent.com/13117724/137129635-779e3c04-bd6e-4e86-a37e-2c4265193ebd.png)
![statusbar](https://user-images.githubusercontent.com/13117724/137129653-7e980708-fc4c-4bbe-a6aa-cbabea645be2.png)

**Comment by BDisp** [comment]
Here is a simple demo:

```cs
Application.Init ();
var lv = new ListView (new List<string> { "A", "B", "C" }) {
	Width = 10,
	Height = 10
};

var btnUp = new Button (0, 11, "^");
btnUp.Clicked += () => {
	if (lv.SelectedItem > 0) {
		var item = lv.Source.ToList () [lv.SelectedItem];
		lv.Source.ToList () [lv.SelectedItem] = lv.Source.ToList () [lv.SelectedItem - 1];
		lv.Source.ToList () [lv.SelectedItem - 1] = item;
		lv.SelectedItem--;
		lv.SetChildNeedsDisplay ();
	}
};
var btnDown = new Button (0, 12, "v");
btnDown.Clicked += () => {
	if (lv.SelectedItem < lv.Source.Count - 1) {
		var item = lv.Source.ToList () [lv.SelectedItem];
		lv.Source.ToList () [lv.SelectedItem] = lv.Source.ToList () [lv.SelectedItem + 1];
		lv.Source.ToList () [lv.SelectedItem + 1] = item;
		lv.SelectedItem++;
		lv.SetChildNeedsDisplay ();
	}
};

var win = new Window ();
win.Add (lv);
win.Add (btnUp);
win.Add (btnDown);

Application.Run (win);
Application.Shutdown ();
```
![up-down-list](https://user-images.githubusercontent.com/13117724/137149479-f742ee36-594c-447e-9568-2382483438a8.png)

**Comment by AnderssonPeter** [comment]
Thank you that should do fine!
